src/modules/fw_path_navigation/FixedwingPositionControl.cpp

```diff
@@ -2458,27 +2454,60 @@ FixedwingPositionControl::reset_landing_state()
 	}
 }
 
-float FixedwingPositionControl::compensateTrimThrottleForDensityAndWeight(float throttle_trim, float throttle_min,
-		float throttle_max)
+float FixedwingPositionControl::calculateTrimThrottle(float throttle_min,
+		float throttle_max, float airspeed_sp)
 {
+	// Approximate drag based on airspeed, vehicle weight and roll angle, and map it to a trim throttle taking the current
+	// air density into account.
+
+	// Drag modelling (parasite drag): calculate mapping airspeed-->throttle, assuming a linear relation with different gradient
+	// above and blow trim. This is tunable thorugh FW_ASPD_MIN and FW_ASPD_MAX.
+	float throttle_trim = _param_fw_thr_trim.get(); // throttle required at sea level during standard conditions.
+
+	if (PX4_ISFINITE(airspeed_sp) && _param_fw_thr_aspd_min.get() > FLT_EPSILON
+	    && airspeed_sp < _param_fw_airspd_trim.get()) {
+		throttle_trim = _param_fw_thr_trim.get() - (_param_fw_thr_trim.get() - _param_fw_thr_aspd_min.get()) /
+				(_param_fw_airspd_trim.get() - _param_fw_airspd_min.get()) * (_param_fw_airspd_trim.get() - airspeed_sp);
+
+	} else if (PX4_ISFINITE(airspeed_sp) && _param_fw_thr_aspd_max.get() > FLT_EPSILON
+		   && airspeed_sp > _param_fw_airspd_trim.get()) {
+		throttle_trim = _param_fw_thr_trim.get() + (_param_fw_thr_aspd_max.get() - _param_fw_thr_trim.get()) /
+				(_param_fw_airspd_max.get() - _param_fw_airspd_trim.get()) * (airspeed_sp - _param_fw_airspd_trim.get());
+	}
+
+	// Drag modelling (induced drag): calculate mapping load factor--> throttle, assuming a linear relation that is independent of
+	// airspeed. This is tunable through FW_INDCD_DRG_RTO (ratio of induced drag to total drag at trim conditions).
+	// The assumption further is that FW_THR_TIM is set correctly and reflects the thrust equivalent of the drag at trim, and that
+	// throttle and thrust are linear --> double the load factor requires double trim throttle * FW_INDCD_DRG_RTO.
+
 	float weight_ratio = 1.0f;
 
 	if (_param_weight_base.get() > FLT_EPSILON && _param_weight_gross.get() > FLT_EPSILON) {
 		weight_ratio = math::constrain(_param_weight_gross.get() / _param_weight_base.get(), MIN_WEIGHT_RATIO,
 					       MAX_WEIGHT_RATIO);
 	}
 
+	float load_factor_from_bank_angle = 1.f;
+
+	if (PX4_ISFINITE(_roll)) {
+		load_factor_from_bank_angle = constrain(1.f / cosf(_roll), 1.f, 2.f);
```

Comment:
@sfuhrer When I run this line on my host machine and I set roll to pi / 2 then the load factor evaluates to `inf` and the `constrain` method returns 2 which I guess is fine.
However, if it returns `NAN` then the functions returns `NAN` which obviously is bad.

Reply:
@sfuhrer Also could you explain the choice of 2 for maximum load factor (45 roll angle) and maybe put it in a const somewhere?

Reply:
No longer relevant as I removed this change again.